Plutarch/DataRepr/Field.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.DataRepr.Field (
+  -- * PDataField class & deriving utils
+  PDataFields (..),
+  pletFields,
+  pletNFields,
+  pletDropFields,
+  pletRangeFields,
+  pfield,
+
+  -- * BindFields class mechanism
+  BindFields (..),
+  type TermsOf,
+  type Take,
+  type Drop,
+
+  -- * Re-exports
+  HRec (..),
+  hrecField,
+) where
+
+import Data.Proxy (Proxy (Proxy))
+import GHC.TypeLits (KnownNat)
+
+import Plutarch.Builtin (
+  PAsData,
+  PData,
+  PIsData (pfromData),
+  pasConstr,
+  psndBuiltin,
+ )
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PIsDataRepr (type PIsDataReprRepr),
+  PIsDataReprInstances,
+  PLabeledType ((:=)),
+  pdhead,
+  pdropDataRecord,
+  pdtail,
+  pindexDataRecord,
+  type PLabel,
+  type PUnLabel,
+ )
+import Plutarch.DataRepr.Internal.HList (
+  HList (HCons, HNil),
+  HRec (HRec),
+  hrecField,
+  type Drop,
+  type IndexList,
+  type IndexOf,
+  type Range,
+  type SingleItem,
+  type Take,
+ )
+import Plutarch.Internal (TermCont (TermCont, runTermCont), punsafeCoerce)
+import Plutarch.Prelude
+
+--------------------------------------------------------------------------------
+---------- PDataField class & deriving utils
+
+{- |
+  Class allowing 'letFields' to work for a PType, usually via
+  `PIsDataRepr`, but is derived for some other types for convenience.
+-}
+class PDataFields (a :: PType) where
+  -- | Fields in HRec bound by 'letFields'
+  type PFields a :: [PLabeledType]
+
+  -- | Convert a Term to a 'PDataList'
+  ptoFields :: Term s a -> Term s (PDataRecord (PFields a))
+
+instance PDataFields (PDataRecord as) where
+  type PFields (PDataRecord as) = as
+  ptoFields = id
+
+instance
+  forall a.
+  ( PIsDataRepr a
+  ) =>
+  PDataFields (PIsDataReprInstances a)
+  where
+  type
+    PFields (PIsDataReprInstances a) =
+      SingleItem (PIsDataReprRepr a)
+
+  ptoFields t =
+    (punsafeCoerce $ phoistAcyclic $ plam $ \d -> psndBuiltin #$ pasConstr # d)
+      # (punsafeCoerce t :: Term _ PData)
+
+instance
+  forall a.
+  ( PIsData a
+  , PDataFields a
+  ) =>
+  PDataFields (PAsData a)
+  where
+  type PFields (PAsData a) = PFields a
+  ptoFields = ptoFields . pfromData
+
+{- |
+  Bind a HRec of named fields from a compatible type.
+-}
+pletFields ::
+  forall a b s.
+  ( PDataFields a
+  , BindFields (PFields a)
+  ) =>
+  Term s a ->
+  (HRec (PLabel (PFields a)) (TermsOf s (PUnLabel (PFields a))) -> Term s b) ->
```

Comment:
The reason why I recommended merging the type parameters is that this type is a bit scary for users.

Reply:
Yeah, and I think it would make it easier to add good type-errors